src/EditorFeatures/CSharp/Highlighting/KeywordHighlighters/AsyncAwaitHighlighter.cs

```diff
@@ -26,93 +30,90 @@ protected override bool IsHighlightableNode(SyntaxNode node)
-                        if (index >= 0)
-                        {
-                            var span = spans[index];
-                            spans[index] = TextSpan.FromBounds(span.Start, awaitToken.Span.End);
-                            handled = true;
-                        }
-                    }
+            using var pooledObject = s_stackPool.GetPooledObject();
 
-                    if (!handled)
-                    {
-                        spans.Add(awaitToken.Span);
-                    }
-                    break;
+            var stack = pooledObject.Object;
+            stack.Push(node);
 
-                case UsingStatementSyntax usingStatement:
-                    if (usingStatement.AwaitKeyword.Kind() == SyntaxKind.AwaitKeyword)
-                    {
-                        spans.Add(usingStatement.AwaitKeyword.Span);
-                    }
-                    break;
+            while (stack.Count > 0)
+            {
+                var current = stack.Pop();
+                yield return current;
 
-                case LocalDeclarationStatementSyntax localDeclaration:
-                    if (localDeclaration.AwaitKeyword.Kind() == SyntaxKind.AwaitKeyword && localDeclaration.UsingKeyword.Kind() == SyntaxKind.UsingKeyword)
+                foreach (var child in current.ChildNodes())
+                {
+                    // Only process children if they're not the start of another construct
+                    // that async/await would be related to.
+                    if (!child.IsReturnableConstruct())
                     {
-                        spans.Add(localDeclaration.AwaitKeyword.Span);
+                        stack.Push(child);
                     }
-                    break;
+                }
+            }
+        }
 
-                case CommonForEachStatementSyntax forEachStatement:
-                    if (forEachStatement.AwaitKeyword.Kind() == SyntaxKind.AwaitKeyword)
-                    {
-                        spans.Add(forEachStatement.AwaitKeyword.Span);
-                    }
-                    break;
+        private static bool HighlightRelatedKeywords(SyntaxNode node, List<TextSpan> spans)
+            => node switch
+            {
+                MethodDeclarationSyntax methodDeclaration => TryAddAsyncModifier(methodDeclaration.Modifiers, spans),
+                LocalFunctionStatementSyntax localFunction => TryAddAsyncModifier(localFunction.Modifiers, spans),
+                AnonymousFunctionExpressionSyntax anonymousFunction => TryAddAsyncOrAwaitKeyword(anonymousFunction.AsyncKeyword, spans),
+                UsingStatementSyntax usingStatement => TryAddAsyncOrAwaitKeyword(usingStatement.AwaitKeyword, spans),
+                LocalDeclarationStatementSyntax localDeclaration =>
+                    localDeclaration.UsingKeyword.Kind() == SyntaxKind.UsingKeyword && TryAddAsyncOrAwaitKeyword(localDeclaration.AwaitKeyword, spans),
```

Comment:
i preserved this logic. But i dont' actually think it's necessary. this basically ensures that we only highlight await in await using var whatever and not in await var whatever. However, the latter is broken code, so it would be fine IMO to highlight there.